tests/Extensions/Hosting.Services.Web.UnitTests/ListenerValidator.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Fabric;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Omex.Extensions.Hosting.Services;
+using Microsoft.Omex.Extensions.Hosting.Services.Web;
+using Microsoft.Omex.Extensions.Logging;
+using Microsoft.Omex.Extensions.TimedScopes;
+using Microsoft.ServiceFabric.Services.Communication.AspNetCore;
+using Microsoft.ServiceFabric.Services.Communication.Runtime;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+
+namespace Hosting.Services.Web.UnitTests
+{
+	internal class ListenerValidator
+	{
+		public ListenerValidator()
+		{
+			m_logsEventSourcMock = new Mock<ILogEventSender>();
+			ListenerName = "TestListener";
+			IntegrationOptions = ServiceFabricIntegrationOptions.None;
+			BuilderAction = builder =>
+			{
+				builder.ConfigureServices(collection =>
+				{
+					collection
+						.AddTransient<TypeRegistredInListenerExtension>()
+						.AddSingleton(m_logsEventSourcMock.Object);
+				});
+			};
+		}
+
+
+		public string ListenerName { get; }
+		public ServiceFabricIntegrationOptions IntegrationOptions { get; }
+		public Action<IWebHostBuilder> BuilderAction { get; }
+
+
+		public IWebHost ValidateListenerBuilder<TContext>(TContext context, KestrelListenerBuilder<Startup, TContext> builder)
+			where TContext : ServiceContext
+		{
+			Assert.AreEqual(ListenerName, builder.Name);
+
+			IWebHost host = builder.BuildWebHost(
+					context,
+					string.Empty,
+					new MockListener(context, (s, l) => new Mock<IWebHost>().Object));
+
+			Assert.ReferenceEquals(context, ResolveType<ServiceContext>(host));
+			Assert.ReferenceEquals(context, ResolveType<TContext>(host));
+			Assert.ReferenceEquals(context, ResolveType<IServiceContextAccessor<TContext>>(host).ServiceContext);
+
+
+			ResolveType<TypeRegistredInListenerExtension>(host);
+			ResolveType<TypeRegistredInStartup>(host);
+
+			return host;
+		}
+
+
+		public void ValidateOmexTypesRegistred(IWebHost host)
+		{
+			ResolveType<ITimedScopeProvider>(host);
+			ILogger logger = ResolveType<ILogger<ListenerValidator>>(host);
+			m_logsEventSourcMock.Invocations.Clear();
+			logger.LogError("TestMessage");
+			Assert.AreNotEqual(0, m_logsEventSourcMock.Invocations.Count, "Omex logger should be registred in WebHost");
+		}
+
+
+		public ICommunicationListener ValidateBuildFunction<TContext>(TContext context, KestrelListenerBuilder<Startup, TContext> builder)
+			where TContext : ServiceContext
+		{
+			ICommunicationListener listener = builder.Build(context);
+			Assert.IsNotNull(listener, "Listener should not be null");
+			return listener;
+		}
+
+
+		private Mock<ILogEventSender> m_logsEventSourcMock;
+
+
+		private T ResolveType<T>(IWebHost host) where T : class
```

Comment:
Use full name for `T`

Reply:
To be honest I don't know what name I can use since method build to resolve any type. 
I'm fine with renaming if you can suggest something, but otherwise, I think it's fine to leave it as is since it's a private method in unit tests class.

Reply:
I'd suggest `TType` or `TOriginalType` in that case. `T` is usually considered too generic, so something more specific is preferable, even if it's just to say "this is very generic" :)

Reply:
I don't think it will add any clarity here `T` already means Type

Reply:
My biggest concern with `T` is that it looks like the default was just used, which then suggests the type could be really anything. Making it more specific indicates that it is generic by design and will work for a wide variety of objects.

Reply:
I understand but it's like making our code worse because other people using `T` wrong. Also, I,m not a strong believer that the name of the generic parameter will stop method user from putting anything in it :) that's the task of generic constraints, names are more for convenience of the person why creates methods.
Quoting code guide:
```
Do name generic type parameters with descriptive names, unless a single letter name is completely self explanatory and a descriptive name would not add value.
```
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/generic-type-parameters